flips/20211129-cadence-mutability-restrictions.md

```diff
@@ -0,0 +1,418 @@
+
+```
+pub struct Foo {
+    pub let x: [Int]
+
+    init() {
+        self.x = [3]
+    }
+}
+
+pub fun bar() {
+    let foo = Foo()
+    foo.x = [0] // writes to x, not allowed
+    foo.x[0] = 0 // does not write to x, currently allowed
+}
+```
+
+This FLIP would change Cadence to also restrict the scopes in which an array or dictionary field can be modified (or "mutated"). 
+Examples of mutating operations include an indexed assignment, as in the above example, as well as calls to the `append` or `remove`,
+methods of arrays, or the `insert` or `remove` methods of dictionaries. These operations will only be able to be performed on a field
+in the current and inner scopes, the same contexts in which the field could be written to if it were a `var`. So the following 
+would typecheck:
+
+```
+pub struct Foo {
+    pub let x : [Int]
+
+    init() {
+        self.x = [3];
+    }
+
+    pub fun addToX(i: Int) {
+        self.x.append(i)
+    }
+}
+```
+
+while the following would not:
+
+```
+pub struct Foo {
+    pub let y : [Int]
+
+    init() {
+        self.y = [3];
+    }
+}
+
+pub fun addToY(foo: Foo, i: Int) {
+    foo.y.append(i)
+}
+```
+
+This is designed to prevent external code from mutating the values of fields it can read from your contract. Consumers
+of your contract may read the values in a `pub let` or `pub var` field, but cannot change them in any way. 
+
+If you wish to allow other code to update or modify a field in your contract, you may expose a method 
+to do so, like in the example above with `addToX`, or you may use the `pub(set)` access mode, which 
+allows any code to mutate or write to the field it applies to. 
+
+### Implementation Details 
+
+This change adds a new error, the `ExternalMutationError`, which is raised when a field 
+is mutated outside of the context in which it was defined. The error message will also
+suggest that the user instead use a setter or modifier method for that field.
+
+Specifically, the error is reported whenever a user attempts to perform an 
+index assignment on a member that is not either declared with the `pub(set)` 
+access mode, or is defined in the current enclosing scope. This check is the
+same one performed for writing to fields, with the difference that mutation 
+is allowed on both `let` and `var` fields, while only the latter can be written to.
+
+Additionally, array and dictionary methods now track an additional bit of information
+indicating whether they mutate their receiver. Mutating methods may not be called on 
+members that are not declared with the `pub(set)` access mode, or defined in the current
+enclosing scope. 
+
+The array methods that are considered mutating are `append`, `appendAll`, `remove`,
+`insert`, `removeFirst` and `removeLast`. The dictionary methods that are considered
+mutating are `insert` and `remove`. Additionally, indexed assignment (`x[y] = e`) will
+be considered mutating for both arrays and dictionaries.  
```

Comment:
```suggestion
mutating are `insert` and `remove`. Additionally, indexed assignment (e.g. `x[y] = e`) will
be considered mutating for both arrays and dictionaries.  
```

Maybe mention that this will apply to all assignment operators, e.g. also the swap operator (`<->`)